beacon-chain/core/blocks/payload.go

```diff
@@ -0,0 +1,140 @@
+package blocks
+
+import (
+	"bytes"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	fieldparams "github.com/prysmaticlabs/prysm/config/fieldparams"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+)
+
+// MergeComplete returns true if the transition to merge has completed.
+// Meaning the header header in beacon state is not `ExecutionPayloadHeader()` (i.e. not empty).
+//
+// Spec code:
+// def is_merge_complete(state: BeaconState) -> bool:
+//    return state.latest_execution_payload_header != ExecutionPayloadHeader()
+func MergeComplete(st state.BeaconState) (bool, error) {
+	h, err := st.LatestExecutionPayloadHeader()
+	if err != nil {
+		return false, err
+	}
+
+	return !isEmptyHeader(h), nil
+}
+
+// MergeBlock returns true if the input block is the terminal merge block.
+// Meaning the header header in beacon state is  `ExecutionPayloadHeader()` (i.e. empty).
+// And the input block has a non-empty header.
+//
+// Spec code:
+// def is_merge_block(state: BeaconState, body: BeaconBlockBody) -> bool:
+//    return not is_merge_complete(state) and body.execution_payload != ExecutionPayload()
+func MergeBlock(st state.BeaconState, blk block.BeaconBlockBody) (bool, error) {
+	mergeComplete, err := MergeComplete(st)
+	if err != nil {
+		return false, err
+	}
+	if mergeComplete {
+		return false, err
+	}
+
+	payload, err := blk.ExecutionPayload()
+	if err != nil {
+		return false, err
+	}
+	return !isEmptyPayload(payload), nil
+}
+
+func isEmptyPayload(p *ethpb.ExecutionPayload) bool {
+	if !bytes.Equal(p.ParentHash, make([]byte, fieldparams.RootLength)) {
+		return false
+	}
+	if !bytes.Equal(p.FeeRecipient, make([]byte, fieldparams.FeeRecipientLength)) {
```

Comment:
This field is still called coinbase in this branch proto's.

Reply:
> This field is still called coinbase in this branch proto's.

The situation with proto's is really a labyrinth, it is `coinbase` in `proto/eth/v2/beacon_block.proto` but it is `fee_recipient` in `proto/prysm/v1alpha1/beacon_block.proto`.

Reply:
`proto/eth/v2` is only used for ethapi. Updating it is out of the scope for this PR. We can do it in another one